File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-18.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        result = float('inf')

        left = 0
        right = len(nums) - 1

        while left <= right:

            if nums[left] < nums[right]:
                result = min(result, nums[left])

            middle = (left + right) // 2

            result = min(nums[middle], result)

            if nums[middle] >= nums[left]:
                left = middle + 1
            else:
                right = middle - 1
        
        return result

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-18.java
class Solution {
    public int characterReplacement(String s, int k) {
        //most repeated = X -> 1, Y -> 2
        //0 1 2 3
        //X Y Y X  K = 2
        //l
        //    r

        Map<Character, Integer> values = new HashMap<>();

        int left = 0;
        int right = 0;
        int mostRepeated = 0;
        int result = 0; 

        while(right < s.length()){
            
            char current = s.charAt(right);
            values.put(current, values.getOrDefault(current, 0) + 1);

            mostRepeated = Math.max(values.get(current), mostRepeated);


            while((right - left + 1) - mostRepeated > k){
                char leftValue = s.charAt(left);
                values.put(leftValue, values.get(leftValue) -  1);
                left++;
            }

            result = Math.max((right - left + 1), result);
            right++;
        }

        return result;
    }
}
